library/ssl_tls13_client.c

```diff
@@ -595,6 +595,320 @@ static int ssl_tls13_write_cookie_ext( mbedtls_ssl_context *ssl,
+    return ( 0 );
+}
+#endif /* MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED */
+
+/*
+ * mbedtls_ssl_tls13_write_pre_shared_key_ext() structure:
+ *
+ * struct {
+ *   opaque identity<1..2^16-1>;
+ *   uint32 obfuscated_ticket_age;
+ * } PskIdentity;
+ *
+ * opaque PskBinderEntry<32..255>;
+ *
+ * struct {
+ *   select ( Handshake.msg_type ) {
+ *
+ *     case client_hello:
+ *       PskIdentity identities<7..2^16-1>;
+ *       PskBinderEntry binders<33..2^16-1>;
+ *
+ *     case server_hello:
+ *       uint16 selected_identity;
+ *   };
+ *
+ * } PreSharedKeyExtension;
+ *
+ *
+ * part = 0 ==> everything up to the PSK binder list,
+ *              returning the binder list length in `binder_list_length`.
+ * part = 1 ==> the PSK binder list
+ */
+
+#if defined(MBEDTLS_KEY_EXCHANGE_SOME_PSK_ENABLED)
+
+int mbedtls_ssl_tls13_write_pre_shared_key_ext_without_binders(
+    mbedtls_ssl_context *ssl,
+    unsigned char *buf, unsigned char *end,
+    size_t *out_len, size_t *binders_len )
+{
+    unsigned char *p = buf;
+    const unsigned char *psk;
+    size_t psk_len;
+    const unsigned char *psk_identity;
+    size_t psk_identity_len;
+    const mbedtls_ssl_ciphersuite_t *suite_info = NULL;
+    const int *ciphersuites;
+    int hash_len = 0;
+    size_t identities_len, l_binders_len;
+    uint32_t obfuscated_ticket_age = 0;
+    psa_algorithm_t psa_hash_alg;
+
+    *out_len = 0;
+    *binders_len = 0;
+
+    /* Check if we have any PSKs to offer. If so, return the first.
+     *
+     * NOTE: Ultimately, we want to be able to offer multiple PSKs,
+     *       in which case we want to iterate over them here.
+     *
+     * As it stands, however, we only ever offer one, chosen
+     * by the following heuristic:
+     * - If a ticket has been configured, offer the corresponding PSK.
+     * - If no ticket has been configured by an external PSK has been
+     *   configured, offer that.
+     * - Otherwise, skip the PSK extension.
+     */
+
+    if( mbedtls_ssl_get_psk_to_offer( ssl, &psk, &psk_len,
+                                      &psk_identity, &psk_identity_len ) != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "skip pre_shared_key extensions" ) );
+        return( 0 );
+    }
+
+    /*
+     * Ciphersuite list
+     */
+    ciphersuites = ssl->conf->ciphersuite_list;
+    for ( int i = 0; ciphersuites[i] != 0; i++ )
+    {
+        suite_info = mbedtls_ssl_ciphersuite_from_id( ciphersuites[i] );
+
+        if( suite_info == NULL )
+            continue;
+
+        /* In this implementation we only add one pre-shared-key extension. */
+        ssl->session_negotiate->ciphersuite = ciphersuites[i];
+        ssl->handshake->ciphersuite_info = suite_info;
+        break;
+    }
+
+    if( suite_info != NULL )
+    {
+        psa_hash_alg = mbedtls_psa_translate_md( suite_info->mac );
+        hash_len = PSA_HASH_LENGTH( psa_hash_alg );
+    }
+    if( hash_len == -1 )
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
```

Comment:
This code might cause compatible issues. That is found in #6079.

I think just choose one from SHA256 or SHA384. That is enough.

Reply:
https://github.com/tlswg/tls13-spec/issues/1227 , @xkqian and @ronald-cron-arm . It looks like an spec issue.
thanks @paul-elliott-arm

Reply:
Does that mean we should abandon the ciphersuites check and **hard code** the parameter of `mbedtls_psa_translate_md `to be `MBEDTLS_MD_SHA256 `or `MBEDTLS_MD_SHA384`?

Reply:
Now the search of a ciphersuite is done only in the case of an external PSK (not a PSK associated to a ticket). We can see that as our way of defining the hash to be used with the external PSK.